File: backend/src/samryetha/markdown.py
"""Markdown → 净化 HTML — 镜像 infrastructure/markdown.ts。

客户端只提交 canonical Markdown；服务端渲染 + 净化后存 body_html。
净化白名单/属性/scheme/链接强制 rel+target 与 TS 端一致；渲染引擎不同，
HTML 输出允许观感级差异（sanitize 语义等价），存量行的 body_html 不重算。
"""

from __future__ import annotations

import re

from markdown_it import MarkdownIt
import nh3

ALLOWED_TAGS = {
    "p", "br", "hr", "strong", "em", "s", "u", "del", "ins",
    "a", "ul", "ol", "li", "blockquote", "code", "pre",
    "h1", "h2", "h3", "h4", "h5", "h6",
    "img", "figure", "figcaption", "table", "thead", "tbody", "tr", "th", "td",
    "span", "div",
}

ALLOWED_ATTRIBUTES = {
    "a": {"href", "title", "target"},
    "img": {"src", "alt", "title", "width", "height"},
    "code": {"class"},
    "span": {"class"},
    "div": {"class"},
    "th": {"align", "colspan"},
    "td": {"align", "colspan"},
}


def _md() -> MarkdownIt:
    md = MarkdownIt("default", {"html": False, "breaks": True, "linkify": False})
    md.enable(["table", "strikethrough"])
    return md


# 为 <a> 强制 target/rel（对应 sanitize-html transformTags）
_ANCHOR_RE = re.compile(r"<a(?=[\s>])", re.IGNORECASE)


def render_markdown(markdown: str) -> str:
    raw = _md().render(markdown or "")
    clean = nh3.clean(
        raw,
        tags=ALLOWED_TAGS,
        attributes=ALLOWED_ATTRIBUTES,
        url_schemes={"http", "https", "mailto"},
    )
    # 为 <a> 强制 rel + target（镜像 sanitize-html transformTags）
    clean = _ANCHOR_RE.sub('<a target="_blank" rel="noopener noreferrer nofollow"', clean)
    return clean

File: backend/src/infrastructure/markdown.ts
import { marked } from "marked";
import sanitizeHtml from "sanitize-html";

// 客户端提交 canonical Markdown，服务端渲染 + 净化后缓存 HTML。
// 绝不直接信任客户端 HTML。
const ALLOWED_TAGS = [
  "p", "br", "hr", "strong", "em", "s", "u", "del", "ins",
  "a", "ul", "ol", "li", "blockquote", "code", "pre",
  "h1", "h2", "h3", "h4", "h5", "h6",
  "img", "figure", "figcaption", "table", "thead", "tbody", "tr", "th", "td",
  "span", "div",
];

const ALLOWED_ATTRIBUTES = {
  a: ["href", "title", "target", "rel"],
  img: ["src", "alt", "title", "width", "height"],
  code: ["class"],
  span: ["class"],
  div: ["class"],
  th: ["align", "colspan"],
  td: ["align", "colspan"],
};

export function renderMarkdown(markdown: string): string {
  const raw = marked.parse(markdown, { async: false, gfm: true, breaks: true }) as string;
  return sanitizeHtml(raw, {
    allowedTags: ALLOWED_TAGS,
    allowedAttributes: ALLOWED_ATTRIBUTES,
    allowedSchemes: ["http", "https", "mailto"],
    allowProtocolRelative: false,
    disallowedTagsMode: "discard",
    transformTags: {
      a: sanitizeHtml.simpleTransform("a", { rel: "noopener noreferrer nofollow", target: "_blank" }),
    },
  });
}
